state/state.go

```diff
@@ -27,3 +36,49 @@ type View interface {
 	NewView(ctx context.Context, changes merkledb.ViewChanges) (merkledb.View, error)
 	GetMerkleRoot(ctx context.Context) (ids.ID, error)
 }
+
+// StateKey holds the name of the key and its permission (Read/Write). By default,
+// initialization of Keys with duplicate key will not work. And to prevent duplicate
+// insertions from overriding the original permissions, use the Add function below.
+// Permissions uses bits since transactions are expected to include this encoding
+type Keys map[string]Permissions
+
+// The access permission (Read/Write) for each StateKey
+// Specifying RWrite is setting both the Read and Write bit
+type Permissions set.Bits
+
+type PermissionBit int
+
+// By default, no permission bits are set
+func Permission(permissions ...PermissionBit) Permissions {
+	perms := set.NewBits()
+	for _, v := range permissions {
+		switch {
+		case v < PermissionLen:
+			// Only set bit to 1 if we're within 8 bits
+			perms.Add(int(v))
+		default:
+			continue
+		}
+	}
+	return Permissions(perms)
+}
+
+// Checks whether a StateKey has the appropriate permission
+// to perform a certain access (Read/Write).
+func (p Permissions) HasPermission(permission PermissionBit) bool {
+	return set.Bits(p).Contains(int(permission))
+}
+
+// Transactions are expected to use this to prevent
+// overriding of key permissions
+func (k Keys) Add(name string, permission Permissions) {
+	_, exists := k[name]
+	if !exists {
+		k[name] = permission
+	} else {
+		// Transaction's permissions are the union of all
+		// state keys from both Actions and Auth
+		set.Bits(k[name]).Union(set.Bits(permission))
```

Comment:
Can we write a simple test that ensures this actually gets set? There is some potentially confusing indirection going on here (because `set.Bits` is modified inplace).

Reply:
I wrote some tests here earlier: https://github.com/ava-labs/hypersdk/pull/693/files#diff-31074509c62c06d7c8b41444cf3f7ac6ee9aa9a1079d2f6d699ae0b063427a69R858 

But I will also make it more clear for that line of code (with the earlier comment you suggested)